src/Data/Text/Encoding.hs

```diff
@@ -406,11 +344,199 @@ streamDecodeUtf8With ::
   HasCallStack =>
 #endif
   OnDecodeError -> ByteString -> Decoding
-streamDecodeUtf8With onErr = go mempty
+streamDecodeUtf8With onErr = go empty . streamDecodeUtf8With'
   where
-    go bs1 bs2 = Some txt undecoded (go undecoded)
+    go t res = case res of
+      ThusFar txt _ bs f -> Some (t `append` txt) bs $ go empty . f
+      InvalidWord txt _ word8 f -> go (t `append` txt) . f . onErr desc $ Just word8
+
+    desc = "Data.Text.Internal.Encoding: Invalid UTF-8 stream"
+
+-- | A stream-oriented decoding result of one of three possibilities.
+data StreamDecode w
+  -- | The decoded 'Text' thus far with a possibly incomplete code
+  -- point at the end of the 'ByteString'. The value contains the
+  -- decoded text up to but not including the incomplete code point,
+  -- the position in the 'ByteString' where the code point starts,
+  -- the incomplete code point, and a function that accepts another
+  -- 'ByteString' as a continuation of the previous input.
+  = ThusFar !Text !Int !ByteString (ByteString -> StreamDecode w)
+  -- | An encounter witt an invalid utf-8 'Word8'. The value contains
+  -- the decoded text up to but not including the invalid 'Word8',
+  -- the position in the 'ByteString' of the offending 'Word8', the
+  -- offender, and a function that accepts a possible 'Char' as to
+  -- interpret the 'Word8' in this specific occurrence of it.
+  | InvalidWord !Text !Int !w (Maybe Char -> StreamDecode w)
```

Comment:
I deliberately left out the third constructor, because I found it redundant with `Incomplete` when the remainder `ByteString` is empty. Also the continuation tracks the byte position across multiple continuation calls for convenience. Just because the `ByteString` was decoded successfully and completely for a given iteration, doesn't mean the caller has finished feeding it data.

However, the responsibility tracking total `Word8`s processed could be pushed to the caller. So if we're discussing simplifying the return type, I would go with something like:
   ```
   -- | Result of decoding
   data StreamDecode w = StreamDecode
     !(Maybe w)  -- ^ If an invalid code point was encountered
     !Text       -- ^ Longest valid prefix
     !ByteString -- ^ Remaining undecoded data after the possible invalid code point
     !Int        -- ^ Byte position of remaining undecoded data
   ```
So complete success with a given `ByteString` input would be represented with `Nothing` for the invalid code point and an empty `ByteString` suffix.

Reply:
So I wrote decoders without the recursion, and using those I wrote decoders with the recursions. Looking at the two, there doesn't seem to be any benefit to non-recursive over the recursive. There's nothing a caller can do with former that can't be done with the latter. If the caller chooses to ignore the continuation function, it's effectively the same as the non-recursive. Also, as I was writing the test cases, the recursion turned out to be a little more convenient.

Reply:
Thanks for doing this. The way I would compare the two instead is that using the non-recursive one you can still easily implement the recursive one, whereas if you try the other way around, it will allocate an unneeded continuation which won't be optimized out because recursive functions are not inlined.

Reply:
I removed the `decodeUtfXStream` functions, but kept the `chunksDecoderToStream` function and moved it to `Data.Text.Encoding.Types`. So a streaming option still remains available, and is simple to produce. Also, since `DecodeResult` contains the remaining unencoded `ByteString`, I make the chunk decoders take two `ByteString` arguments so the caller doesn't have to append the leftovers of the previous call to the new `ByteString` of the next call.